### ISSUE TITLE
Add logging to MetadataService via SettingsService?

### ISSUE THREAD

**Issue by MasterKale** [MasterKale/SimpleWebAuthn#158]
I've been wanting to add some logging to MetadataService specifically to help RP devs understand what's going on behind the scenes when it's enabled. Figuring out a way to output log statements in a library intended for use in other projects is a challenge, though, as I don't want to be prescriptive about how implementations handle logging just so that my library can report on its inner workings.

With the introduction of `SettingsService` in v4.0.0 I realized I now have an way to empower RP devs to control how MetadataService outputs its log statements. I can add some kind of optional `addLogger(logger: (level, message) => void)` method to it, and then retrieve the logger in MetadataService and use it to output logging.

**Comment by MasterKale** [comment]
Leaving some notes for myself, observations of other popular logging libraries and their logging interfaces:

## `console`
```ts
// lib.dom.d.ts
console.info:  (...data: any[]) => void;
console.warn:  (...data: any[]) => void;
console.error: (...data: any[]) => void;
console.debug: (...data: any[]) => void;
```

## `logging`
```ts
// LoggerFunction: https://github.com/dylang/logging/blob/master/index.d.ts#L14
type LoggerFunction = (...messages: unknown[]) => void;
```

## `winston`
```ts
// LeveledLogMethod: https://github.com/winstonjs/winston/blob/master/index.d.ts#L69
interface LeveledLogMethod {
  (message: string, callback: LogCallback): Logger;
  (message: string, meta: any, callback: LogCallback): Logger;
  (message: string, ...meta: any[]): Logger;
  (message: any): Logger;
  (infoObject: object): Logger;
}
```

## `pino`
```ts
// LogFn: https://github.com/pinojs/pino/blob/71a537c285e907005d756f5092f2d9b2b41f142e/pino.d.ts#L886
interface LogFn {
  /* tslint:disable:no-unnecessary-generics */
  <T extends object>(obj: T, msg?: string, ...args: any[]): void;
  (msg: string, ...args: any[]): void;
}
```

## `bunyan`
```ts
// @types/bunyan (each had a few overrides, just grabbed the ones most like the ones above)
info(obj: Object, ...params: any[]): void;
warn(obj: Object, ...params: any[]): void;
error(obj: Object, ...params: any[]): void;
debug(obj: Object, ...params: any[]): void;
```

**Comment by P4sca1** [comment]
What do you think about https://github.com/visionmedia/debug? It is used in quite a few libraries.

When you want to get logs about SimpleWebAuthn, you simply start your app with the env variable `DEBUG=SimpleWebAuthn:*` (or whatever namespace you choose).
You can even have different namespaces, e.g. to debug only the `MetadataService`, one could do `DEBUG=SimpleWebAuthn:MetadataService`.

**Comment by MasterKale** [comment]
`debug` was one of the libraries I considered. In PR #97 I decided to try and incorporate [anylogger](https://www.npmjs.com/package/anylogger) instead because it offered an interface for other logging libraries to tap into the logs coming out of MetadataService, including debug via [anylogger-debug](https://www.npmjs.com/package/anylogger-debug).

Unfortunately due to an `npm` issue discovered by the author of anylogger in response to an issue I logged (see https://github.com/Download/anylogger/issues/17#issuecomment-812866038) I don't think I'll ever be able to incorporate anylogger after all because of an idiosyncracy about how `npm` requires dependencies.

That leaves me prescribing how a consumer of SimpleWebAuthn is to enable logging by picking something specific like `debug` and that kinda makes me sad. Hence why I created this ticket to try and identify _some_ way to enable logging without imposing the use of specific env vars/etc... just for this library.

**Comment by P4sca1** [comment]
If I understood correctly, enabling logging for a library that uses anylogger involves the following:
1. Install anylogger and an anylog adapter for your current logging solution
2. Require the anylog adapter and hope that anylogger in the library you use detects that you have required it in your application. This approach seems very unstable and requires great care in the build process. IMO it’s not even a problem with npm. Relying on the fact, that there is ever only one version of the Package across all dependencies is bad. Imagine one other library you use, requires an older version of anylogger than another. This can easily happen and seems to break the whole concept of anylogger or am I wrong?

As a library consumer, the debug approach is very easy. I don’t have to install any dependencies. Just set that env variable and you are good to go. You can even integrate debug into your existing logging solution (similar to anylogger). See pino-debug for example.

**Comment by P4sca1** [comment]
And as there are many libraries already using the debug package, it is likely that one already worked with the DEBUG env variable.

**Comment by MasterKale** [comment]
@P4sca1 I created PR #159 to integrate `debug` for logging. I'd love to get your thoughts on whether it'd make sense to log more of the internals of the library, or if it should only be limited to MetadataService and may SettingsService.

**Comment by P4sca1** [comment]
I would say it's never bad to add more logging later, because you can easily filter it using debug scopes.

**Comment by MasterKale** [comment]
PR #159 has been merged. The new logging is available in the newly-published @simplewebauthn/server@4.2.0.

Thank you @P4sca1 for knocking some sense into me about debug 😂
